src/mumble/pluginAPI/PluginComponents.h

```diff
@@ -0,0 +1,134 @@
+/**
+ * This header file contains definitions of types and other components used in Mumble's plugin system
+ */
+
+#ifndef MUMBLE_PLUGINCOMPONENT_H_
+#define MUMBLE_PLUGINCOMPONENT_H_
+
+#include <stdint.h>
+#include <stddef.h>
+
+#define STATUS_OK 0
+
+enum TalkingState {
+	INVALID=-1,
+	PASSIVE=0,
+	TALKING,
+	WHISPERING,
+	SHOUTING
+};
+
+struct Version {
+	int32_t major;
+	int32_t minor;
+	int32_t patch;
+};
+
+
+typedef enum TalkingState TalkingState_t;
+typedef struct Version Version_t;
+typedef int32_t MumbleConnection_t;
+typedef uint32_t MumbleUserID_t;
+typedef int32_t MumbleChannelID_t;
+typedef int32_t error_t;
+
+
+struct MumbleAPI {
+	// -------- Memory management --------
+	
+	/// Frees the given pointer.
+	///
+	/// @param pointer The pointer to free
+	/// @returns The error code. If everything went well, STATUS_OK will be returned.
+	error_t (*freeMemory)(void *pointer);
+
+
+	
+	// -------- Getter functions --------
+
+	/// Gets the connection ID of the server the user is currently active on (the user's audio output is directed at).
+	///
+	/// @param[out] connection A pointer to the memory location the ID should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then it is valid to access the
+	/// 	value of the provided pointer
+	error_t (*getActiveServerConnection)(MumbleConnection_t *connection);
+
+	/// Fills in the information about the local user.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param[out] userID A pointer to the memory the user's ID shall be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getLocalUserID)(MumbleConnection_t connection, MumbleUserID_t *userID);
+
+	/// Fills in the information about the given user's name. The name allocated if this function returns STATUS_OK
+	/// has to be freed by a call to freeMemory eventually.
+	///
+	/// @param connection The ID of the server-connection to use as a context
+	/// @param userID The user's ID whose name should be obtained
+	/// @param[out] userName A pointer to where the pointer to the allocated string (C-ecoded) should be written to
+	/// @returns The error code. If everything went well, STATUS_OK will be returned. Only then the passed pointer
+	/// 	may be accessed
+	error_t (*getUserName)(MumbleConnection_t connection, MumbleUserID_t userID, const char **userName);
+
+	/// Fills in the information about the given channel's name. The name allocated if this function returns STATUS_OK
+	/// has to be freed by a call to freeMemory eventually.
```

Comment:
Maybe move that note directly onto the parameter that actually has to be freed.
Not sure if that's easier to understand for the user? also we'd then repeat ourselves on API functions that return multiple things

Reply:
Yeah, that might be more clear as to which variable needs to be freed. As for the repetitions I think that this isn't a deal-breaker. After all this is pretty crucial anyways so it doesn't hurt to repeat oneself on that.